From 070d72477df2a41bc085d4ce00c854250dd9ed43 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Tue, 5 Aug 2025 15:47:45 +0200
Subject: [PATCH] phy: stm32: usb2phy: adopt one cell as multiple clocks
 provider

Use one clock cell, as USB2PHY provides several clocks. It allows to
properly manage OHCI clocks while the controller is suspended.
When the OHCI controller is put in suspend state, CMN bit can be set to
one (so the clocks are off). Upon resume, upon clock enabling, CMN bit
must be cleared, in order to properly access the registers.

The PLL (clock index 0) is defined as parent clock of OHCI clocks (clock
index 1).

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: Id86c8e682301745eee180f2ac50328e05b2d6c1e
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/424747
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Christian BRUEL <christian.bruel@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/phy/st/phy-stm32-usb2phy.c | 188 ++++++++++++++++-------------
 1 file changed, 107 insertions(+), 81 deletions(-)

diff --git a/drivers/phy/st/phy-stm32-usb2phy.c b/drivers/phy/st/phy-stm32-usb2phy.c
index 860694c1d7ce..27068d937325 100644
--- a/drivers/phy/st/phy-stm32-usb2phy.c
+++ b/drivers/phy/st/phy-stm32-usb2phy.c
@@ -73,6 +73,8 @@
 #define SYSCFG_USB2PHYTRIM2_TXPREEMPAMPTUNE_MASK	GENMASK(1, 0)
 #define SYSCFG_USB2PHYTRIM2_TXPREEMPPULSETUNE_MASK	BIT(2)
 
+#define USB2PHY_CLK_NUM		2
+
 struct stm32_usb2phy {
 	struct phy *phy;
 	struct regmap *regmap;
@@ -85,7 +87,8 @@ struct stm32_usb2phy {
 	int submode;
 	u32 mask_trim1, value_trim1, mask_trim2, value_trim2;
 	bool is_init;
-	struct clk_hw clk48_hw;
+	struct clk_hw clkpll_hw;
+	struct clk_hw clkohci_hw;
 	atomic_t en_refcnt;
 	const struct stm32mp2_usb2phy_hw_data *hw_data;
 	bool do_wakeup;
@@ -252,21 +255,6 @@ static int stm32_usb2phy_enable(struct stm32_usb2phy *phy_dev)
 		return ret;
 	}
 
-	if (phy_data->valid_mode == USB2_MODE_HOST_ONLY) {
-		/*
-		 * The clock should default to active after standby, as it is
-		 * needed when resuming OHCI to access its registers.
-		 * CMN is default reset to 1, so enforce it is cleared, when the
-		 * clock enable request from OHCI driver comes at resume time.
-		 */
-		ret = regmap_clear_bits(phy_dev->regmap, phy_data->cr_offset,
-					SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK);
-		if (ret) {
-			dev_err(dev, "can't clear CMN bit (%d)\n", ret);
-			return ret;
-		}
-	}
-
 	if (phy_dev->mask_trim1) {
 		ret = regmap_update_bits(phy_dev->regmap, phy_dev->hw_data->trim1_offset,
 					 phy_dev->mask_trim1, phy_dev->value_trim1);
@@ -399,34 +387,25 @@ static int stm32_usb2phy_set_mode(struct phy *phy, enum phy_mode mode, int submo
 	const struct stm32mp2_usb2phy_hw_data *phy_data = phy_dev->hw_data;
 	struct device *dev = &phy->dev;
 
+	/* Simply return when host only or stm32mp21 OTG (controlled by OTG GGPIO) */
+	if (phy_data->valid_mode == USB2_MODE_HOST_ONLY || phy_data->valid_mode == USB2_MODE_OTG)
+		return 0;
+
 	switch (mode) {
 	case PHY_MODE_USB_HOST:
-		if (phy_data->valid_mode == USB2_MODE_HOST_ONLY)
+		if (submode == USB_ROLE_NONE) {
 			ret = regmap_update_bits(phy_dev->regmap,
 						 phy_data->cr_offset,
-						 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK,
-						 0);
-		else if (phy_data->valid_mode == USB2_MODE_OTG)
+						 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVLDEXT_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVALID_MASK, 0);
+		} else {
 			ret = regmap_update_bits(phy_dev->regmap,
 						 phy_data->cr_offset,
-						 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK,
-						 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK);
-		else {
-			if (submode == USB_ROLE_NONE) {
-				ret = regmap_update_bits(phy_dev->regmap,
-							 phy_data->cr_offset,
-							 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVLDEXT_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVALID_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVLDEXT_MASK, 0);
-			} else {
-				ret = regmap_update_bits(phy_dev->regmap,
-							 phy_data->cr_offset,
-							 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVLDEXT_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVALID_MASK,
-							 SYSCFG_USB2PHY2CR_VBUSVALID_MASK);
-			}
+						 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVLDEXT_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVALID_MASK,
+						 SYSCFG_USB2PHY2CR_VBUSVALID_MASK);
 		}
 		if (ret) {
 			dev_err(dev, "can't set usb2phycr (%d)\n", ret);
@@ -435,32 +414,25 @@ static int stm32_usb2phy_set_mode(struct phy *phy, enum phy_mode mode, int submo
 		break;
 
 	case PHY_MODE_USB_DEVICE:
-		if (phy_data->valid_mode == USB2_MODE_OTG)
+		if (submode == USB_ROLE_NONE) {
+			ret = regmap_update_bits(phy_dev->regmap,
+						 phy_data->cr_offset,
+						 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVALID_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVLDEXTSEL_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVLDEXT_MASK,
+						 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVLDEXTSEL_MASK);
+		} else {
 			ret = regmap_update_bits(phy_dev->regmap,
 						 phy_data->cr_offset,
-						 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK,
-						 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK);
-		else {
-			if (submode == USB_ROLE_NONE) {
-				ret = regmap_update_bits(phy_dev->regmap,
-							 phy_data->cr_offset,
-							 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVALID_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVLDEXTSEL_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVLDEXT_MASK,
-							 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVLDEXTSEL_MASK);
-			} else {
-				ret = regmap_update_bits(phy_dev->regmap,
-							 phy_data->cr_offset,
-							 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVALID_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVLDEXTSEL_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVLDEXT_MASK,
-							 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVLDEXTSEL_MASK |
-							 SYSCFG_USB2PHY2CR_VBUSVLDEXT_MASK);
-			}
+						 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVALID_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVLDEXTSEL_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVLDEXT_MASK,
+						 SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVLDEXTSEL_MASK |
+						 SYSCFG_USB2PHY2CR_VBUSVLDEXT_MASK);
 		}
 		if (ret) {
 			dev_err(dev, "can't set usb2phycr (%d)\n", ret);
@@ -569,55 +541,109 @@ static const struct phy_ops stm32_usb2phy_data = {
 	.owner = THIS_MODULE,
 };
 
-static int stm32_usb2phy_clk48_prepare(struct clk_hw *hw)
+static int stm32_usb2phy_clkpll_prepare(struct clk_hw *hw)
 {
-	struct stm32_usb2phy *phy_dev = container_of(hw, struct stm32_usb2phy,
-							   clk48_hw);
+	struct stm32_usb2phy *phy_dev = container_of(hw, struct stm32_usb2phy, clkpll_hw);
 
 	return stm32_usb2phy_enable(phy_dev);
 }
 
-static void stm32_usb2phy_clk48_unprepare(struct clk_hw *hw)
+static void stm32_usb2phy_clkpll_unprepare(struct clk_hw *hw)
 {
-	struct stm32_usb2phy *phy_dev = container_of(hw, struct stm32_usb2phy,
-							   clk48_hw);
+	struct stm32_usb2phy *phy_dev = container_of(hw, struct stm32_usb2phy, clkpll_hw);
 
 	stm32_usb2phy_disable(phy_dev);
 }
 
-static unsigned long stm32_usb2phy_clk48_recalc_rate(struct clk_hw *hw,
-							   unsigned long parent_rate)
+static unsigned long stm32_usb2phy_clkpll_recalc_rate(struct clk_hw *hw,
+						      unsigned long parent_rate)
+{
+	return 480000000;
+}
+
+static const struct clk_ops stm32_usb2phy_clkpll_ops = {
+	.prepare = stm32_usb2phy_clkpll_prepare,
+	.unprepare = stm32_usb2phy_clkpll_unprepare,
+	.recalc_rate = stm32_usb2phy_clkpll_recalc_rate,
+};
+
+static int stm32_usb2phy_clkohci_prepare(struct clk_hw *hw)
+{
+	struct stm32_usb2phy *phy_dev = container_of(hw, struct stm32_usb2phy, clkohci_hw);
+	const struct stm32mp2_usb2phy_hw_data *phy_data = phy_dev->hw_data;
+	int ret;
+
+	ret = regmap_clear_bits(phy_dev->regmap, phy_data->cr_offset,
+				SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK);
+
+	return ret;
+}
+
+static void stm32_usb2phy_clkohci_unprepare(struct clk_hw *hw)
+{
+	struct stm32_usb2phy *phy_dev = container_of(hw, struct stm32_usb2phy, clkohci_hw);
+	const struct stm32mp2_usb2phy_hw_data *phy_data = phy_dev->hw_data;
+
+	regmap_set_bits(phy_dev->regmap, phy_data->cr_offset, SYSCFG_USB2PHY2CR_USB2PHY2CMN_MASK);
+}
+
+static unsigned long stm32_usb2phy_clkohci_recalc_rate(struct clk_hw *hw,
+						       unsigned long parent_rate)
 {
 	return 48000000;
 }
 
-static const struct clk_ops stm32_usb2phy_clk48_ops = {
-	.prepare = stm32_usb2phy_clk48_prepare,
-	.unprepare = stm32_usb2phy_clk48_unprepare,
-	.recalc_rate = stm32_usb2phy_clk48_recalc_rate,
+static const struct clk_ops stm32_usb2phy_clkohci_ops = {
+	.prepare = stm32_usb2phy_clkohci_prepare,
+	.unprepare = stm32_usb2phy_clkohci_unprepare,
+	.recalc_rate = stm32_usb2phy_clkohci_recalc_rate,
 };
 
 static int stm32_usb2phy_clk48_register(struct stm32_usb2phy *phy_dev)
 {
+	struct clk_hw_onecell_data *clk_data;
 	struct clk_init_data init = { };
 	int ret = 0;
 	char name[20];
 
-	snprintf(name, sizeof(name), "ck_usb2phy%x_48m", phy_dev->hw_data->cr_offset);
+	snprintf(name, sizeof(name), "ck_usb2phy%x_pll", phy_dev->hw_data->cr_offset);
 	init.name = name;
-	init.ops = &stm32_usb2phy_clk48_ops;
+	init.ops = &stm32_usb2phy_clkpll_ops;
+
+	phy_dev->clkpll_hw.init = &init;
+
+	ret = devm_clk_hw_register(phy_dev->dev, &phy_dev->clkpll_hw);
+	if (ret) {
+		dev_err(phy_dev->dev, "failed to register pll clk\n");
+		return ret;
+	}
+
+	snprintf(name, sizeof(name), "ck_usb2phy%x_48m", phy_dev->hw_data->cr_offset);
 
-	phy_dev->clk48_hw.init = &init;
+	phy_dev->clkohci_hw.init = CLK_HW_INIT_HW(name, &phy_dev->clkpll_hw,
+						  &stm32_usb2phy_clkohci_ops, 0);
 
-	ret = devm_clk_hw_register(phy_dev->dev, &phy_dev->clk48_hw);
+	ret = devm_clk_hw_register(phy_dev->dev, &phy_dev->clkohci_hw);
 	if (ret) {
-		dev_err(phy_dev->dev, "failed to register 48m clk\n");
+		dev_err(phy_dev->dev, "failed to register ohci clk\n");
 		return ret;
 	}
 
-	ret = devm_of_clk_add_hw_provider(phy_dev->dev, of_clk_hw_simple_get, &phy_dev->clk48_hw);
+	/*
+	 * USB2PHY provides several clocks used either by either USHB (EHCI/OHCI), OTG or USB3DR.
+	 * In case of OHCI, CMN bit must be cleared (clkohci_hw). This clock is required to access
+	 * the registers, to resume the controller from suspended state.
+	 * So declare two clocks, the PLL used in all case, and the OHCI clocks used by OHCI
+	 * controller.
+	 */
+	clk_data = devm_kzalloc(phy_dev->dev, struct_size(clk_data, hws, USB2PHY_CLK_NUM),
+				GFP_KERNEL);
+	clk_data->num = USB2PHY_CLK_NUM;
+	clk_data->hws[0] = &phy_dev->clkpll_hw;
+	clk_data->hws[1] = &phy_dev->clkohci_hw;
+	ret = devm_of_clk_add_hw_provider(phy_dev->dev, of_clk_hw_onecell_get, clk_data);
 	if (ret)
-		dev_err(phy_dev->dev, "adding clk provider 48m failed\n");
+		dev_err(phy_dev->dev, "adding clk provider failed\n");
 
 	return ret;
 }
-- 
2.53.0

